### PR TITLE
Switch to sequential mode on long partition names

### PR DESCRIPTION
**Pull request by hanefi**
DESCRIPTION: Removes limits around long partition names

Long partition table names can create distributed deadlocks. If we switch to sequential execution, then we can handle long partition names when creating or attaching partitions to distributed tables.

This approach uses new shard id values to create hypothetical shard names before distributing partitions. A side effect is discarding some shard id values. In other words we will have gaps between shard ids of actual shards.

Fixes: #4736
Closes: #4793 
Related: #4739

### CODE REVIEW

File: src/test/regress/sql/multi_name_lengths.sql [hunk 1]
```diff
@@ -130,14 +130,9 @@ ALTER TABLE partition_lengths RENAME TO partition_lengths_1234567890123456789012
...
--- Placeholders for unsupported operations
-\set VERBOSITY TERSE
-
--- renaming distributed table partitions
-ALTER TABLE partition_lengths_p2020_09_28 RENAME TO partition_lengths_p2020_09_28_12345678901234567890123456789012345678901234567890;
 
--- creating or attaching new partitions with long names create deadlocks
-CREATE TABLE partition_lengths_p2020_09_29_12345678901234567890123456789012345678901234567890 (LIKE partition_lengths_p2020_09_28_12345678901234567890123456789012345678901234567890);
+-- creating or attaching new partitions with long names
+CREATE TABLE partition_lengths_p2020_09_29_12345678901234567890123456789012345678901234567890 (LIKE partition_lengths_p2020_09_28);
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/test/regress/sql/multi_name_lengths.sql:135]: We already have a test several lines below. I do not see the value in adding yet another one.

Nevertheless, I added one

File: src/test/regress/sql/multi_name_lengths.sql [hunk 2]
```diff
@@ -215,8 +207,19 @@ SELECT master_create_distributed_table('sneaky_name_lengths', 'int_col_123456789
 SELECT master_create_worker_shards('sneaky_name_lengths', '2', '2');
 
 \c - - :public_worker_1_host :worker_1_port
-\di public.sneaky*225030
```
Comment (hanefi) [inline src/test/regress/sql/multi_name_lengths.sql:218]: The hardcoded shardId here makes it problematic to update some of the tests above. That's why I am doing a lookup to find an index name on a shard.

I could have also write a query that is similar to `/di <index_name>`. However I think this is ok for now.

File: src/backend/distributed/commands/table.c [hunk 3]
```diff
@@ -324,6 +324,9 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
 		char *parentRelationName = generate_qualified_relation_name(parentRelationId);
 		bool viaDeprecatedAPI = false;
 
+		SwitchToSequentialAndLocalExecutionIfRelationNameTooLong(
```
Comment (hanefi) [inline src/backend/distributed/commands/table.c:327]: I don't really like the indentation in here. Sadly, I see no other option to fix this.

File: src/backend/distributed/utils/shard_utils.c [hunk 4]
```diff
@@ -55,3 +56,21 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
+
+/*
+ * GetLongestShardName is a utility function that creates a hypothetical shard
+ * name for a table that is not distributed yet.
+ *
+ * Note that this operation will increment the shard id seqeunce without
+ * actually using the shard id in an actual shard.
+ */
+char *
+GetLongestHypotheticalShardName(char *relationName)
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c]: I implemented what you suggested, with a small change.

> shardCount of the parentTable
instead of this, I used the current `citus.shard_count` config value as accessing the shardCount of the parent table will not be as clean and readable IMO.

Also `LoadShardIntervalWithLongestShardName` returns the largest shardId of the parent table. Even if we add 1 to this value, we will hit the shardId of the next table that is distributed after this parent table.

File: src/backend/distributed/utils/shard_utils.c [hunk 5]
```diff
@@ -55,3 +56,23 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
+ * There is no easy way to create the names of shards before those shards are created
+ * because sequence manipulation in PostgreSQL is not transactional. Therefore we assume
+ * that the shardId values of a partition and those of its parent will be adjacent.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
+{
+	char *longestShardName = pstrdup(relationName);
+	ShardInterval *shardInterval = LoadShardIntervalWithLongestShardName(parentTableOid);
+	uint64 newShardId = shardInterval->shardId + ShardCount;
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:74]: Should we use 
```
CitusTableCacheEntry *cache = GetCitusTableCacheEntry(relationId);
cache->shardIntervalArrayLength
```
Instead of `ShardCount`. Because we might be dealing with a table that has a different shard count?

File: src/backend/distributed/utils/shard_utils.c [hunk 6]
```diff
@@ -55,3 +56,23 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
 
 	return longestShardName;
 }
+
+
+/*
+ * GetLongestShardName is a utility function that creates a hypothetical shard
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:62]: minor: GetLongestHypotheticalShardName

File: src/test/regress/sql/multi_name_lengths.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/sql/multi_name_lengths.sql:135]: > We already have a test several lines below.

I didn't see it, feel free to remove if you prefer so

File: src/backend/distributed/utils/shard_utils.c [hunk 7]
```diff
@@ -55,3 +56,23 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
+ * name for a table that is not distributed yet.
+ *
+ * There is no easy way to create the names of shards before those shards are created
+ * because sequence manipulation in PostgreSQL is not transactional. Therefore we assume
+ * that the shardId values of a partition and those of its parent will be adjacent.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:70]: I don't want to over-engineer this, but parent table's shard could become outdated pretty quickly. So, assume we create 1 partition per day with shardCount 128, and hence the parent's largest shardId is could be too low after few days.

Ok, we don't want to call `GetNextShardId()`, but can we add `GetCurrentShardId()` and then get parents `` to generate the shardName:

```C
		CitusTableCacheEntry *parentCache = GetCitusTableCacheEntry(parentRelId);
        int shardCount = parentCache->shardIntervalArrayLength;
      
      int maxShardCount  = GetCurrentShardId()  + shardCount;
      char *newPartitionsName = pstrdup(get_rel_name(partitionId));
     return AppendShardIdToName(&newPartitionsName, maxShardCount);
```

Please first verify if this approach works fine, I just come up with this and didn't have much time to verify.

Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c:70]: AFAIK there is no easy way to get the last value generated by a sequence. `currval(regclass)` and `lastval()` gives errors if the `nextval()` function was not used earlier in the sessions.

> currval ( regclass ) → bigintReturns the value most recently obtained by nextval for this sequence in the current session. (An error is reported if nextval has never been called for this sequence in this session.) Because this is returning a session-local value, it gives a predictable answer whether or not other sessions have executed nextval since the current session did.This function requires USAGE or SELECT privilege on the sequence.
>
> lastval() → bigintReturns the value most recently returned by nextval in the current session. This function is identical to currval, except that instead of taking the sequence name as an argument it refers to whichever sequence nextval was most recently applied to in the current session. It is an error to call lastval if nextval has not yet been called in the current session.This function requires USAGE or SELECT privilege on the last used sequence.

I can however, iterate over all the shard values in the metadata, and return `1 + maxShardId` and IMO this would be overengineering.

See https://www.postgresql.org/docs/13/functions-sequence.html for the docs

Comment (onderkalaci) [reply src/backend/distributed/utils/shard_utils.c:70]: Can we use `pg_sequences`?
```SQL
select pg_sequence_last_value('pg_catalog.pg_dist_shardid_seq');
```

Because the current approach feels fragile and/or not very consistent. Parent's shardId is mostly likely not useful in many cases. What do you think?

Comment (marcocitus) [reply src/backend/distributed/utils/shard_utils.c:70]: fwiw, `select last_value from pg_dist_shardid_seq` also gives the latest value, but we should probably just use the underlying C functions.

Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c:70]: `pg_sequence_last_value` returns NULL if the sequence never generated any numbers after the last reset.

I think it is safe to assert the sequence `is_called` flag will be true and it is safe to read its' last_value.

I can use the C function for `pg_sequence_last_value` to make this happen.

Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c:70]: I have 2 problems with using `pg_sequence_last_value` C function.
1. It is not declared in any header, and it uses several static functions in the same file. I don't think it is wise to copy the contents of those functions to our codebase as it is not easy to maintain.
2. It can potentially return `NULL` on several occasions:
  - the sequence is never used before
  - the sequence is never used after a command that set its `is_called` flag to false:
  	- `ALTER SEQUENCE .. RESTART`
  	- `SELECT setval( <regclass>, <bigint> , false )` that can set the `is_called` flag via the third parameter that is optional

```
hanefi=# SELECT * FROM pg_dist_shardid_seq ;
 last_value | log_cnt | is_called
------------+---------+-----------
     102008 |       0 | f
(1 row)

hanefi=# SELECT pg_sequence_last_value('pg_catalog.pg_dist_shardid_seq');
 pg_sequence_last_value
------------------------

(1 row)
```

My thinking is that it is better to run a SQL command (`SELECT last_value from pg_dist_shardid_seq`) in here and not use the underlying C functions to make sure the codebase is easier to maintain.

File: src/backend/distributed/utils/shard_utils.c [hunk 8]
```diff
@@ -55,3 +59,52 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
+	PG_TRY();
+	{
+		Datum lastShardIdDatum = DirectFunctionCall1(pg_sequence_last_value,
+													 sequenceIdDatum);
+		newShardId = DatumGetInt64(lastShardIdDatum) + shardIntervalCount;
+	}
+	PG_CATCH();
+	{
+		/* assume that we have a shardId with 7 digits */
+		newShardId = 9999999;
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:101]: We should maybe mark `newShardId` as `volatile`, as the comment suggests in `PG_TRY()` definition:
```
 * Note: if a local variable of the function containing PG_TRY is modified
 * in the PG_TRY section and used in the PG_CATCH section, that variable
 * must be declared "volatile" for POSIX compliance.  This is not mere
 * pedantry; we have seen bugs from compilers improperly optimizing code
 * away when such a variable was not marked.  Beware that gcc's -Wclobbered
 * warnings are just about entirely useless for catching such oversights.
```

We are not using `newShardId` but lets be robust to future changes

File: src/backend/distributed/utils/shard_utils.c [hunk 9]
```diff
@@ -55,3 +59,52 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
 	return longestShardName;
 }
+
+
+/*
+ * GetLongestHypotheticalShardName is a utility function that creates a hypothetical shard
+ * name for a table that is not distributed yet.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:69]: It is kind of hard to follow the naming. The function name is very generic but:

1) We name the function as `GetLongestHypotheticalShardName` implying that we are calling this on a local table. However, the function assumes that the input `parentTableOid` is a distributed table
2) The function parameter is `parentTableOid`, but the function name seems to operate on all kinds of tables

In general, I was thinking that we should probably extract `pg_sequence_last_value` related parts to a different function, and the calculations regarding to parent -> child relations in a different function. This function seems to do many things at once, and slightly implicitly

File: src/backend/distributed/utils/shard_utils.c [hunk 10]
```diff
@@ -55,3 +59,52 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
...
+	PG_CATCH();
+	{
+		/* assume that we have a shardId with 7 digits */
+		newShardId = 9999999;
+	}
+	PG_END_TRY();
+
+	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
+
+	AppendShardIdToName(&longestShardName, newShardId);
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shard_utils.c:107]: should we make `newShardId + 1`?

**Review by onderkalaci — COMMENTED** [review]
Minor code readability notes

**Comment by onderkalaci** [comment]
It is also pretty unexpected user experience that we don't support this properly for regular distributed tables, marking as a related #4739. I guess it is much more common to have longer relation names for partitioned tables, so it is a fair trade-off for now.

File: src/backend/distributed/utils/shard_utils.c (hunk 9 above)
Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c]: 1. How does `GetLongestShardNameForLocalPartition` sound? I renamed it.
2. Would you want me to add an assertion or some error messages that should never occur here? I could not find a better alternative for this.

I also moved `pg_sequence_last_value` calls in a separate helper function

File: src/backend/distributed/utils/shard_utils.c (hunk 10 above)
Comment (hanefi) [reply src/backend/distributed/utils/shard_utils.c]: I don't think this is necessary. Since the value returned by `pg_sequence_last_value` will be the last used shard id value, the next consecutive shard id values should be used by this new partitions.

**Comment by hanefi** [comment]
> It is also pretty unexpected user experience that we don't support this properly for regular distributed tables, marking as a related #4739.

To remedy the issues in regular distributed tables we need to call `SwitchToSequentialAndLocalExecutionIf..` functions in `CreateDistributedTable` calls. I tried to understand how easy it was during this task and it was not trivial.

**Review by onderkalaci — APPROVED** [review]
All looks good.

Again sharing for reference, we should fix https://github.com/citusdata/citus/issues/4739 as well, it is weird that we support ALTER TABLE .. ATTACH partition, but not `create_distributed_table` with the same table name
